### PR TITLE
Use C++20 with more packages

### PR DESCRIPTION
**Pull request by iarspider**

### CODE REVIEW

**Comment by iarspider** [comment]
please test for CMSSW_13_2_CPP20_X

File: professor2-ppc64-flag-change.patch [hunk 1]
```diff
@@ -20,7 +20,7 @@ index adf3b19..fabdbf0 100644
                  depends=glob("../include/*.h"),
                  include_dirs=[incdir, os.path.join(srcdir, "pyext", "professor2")],
 -                extra_compile_args="-std=c++11 -O3 -Wno-unused-but-set-variable -Wno-sign-compare".split(),
-+                extra_compile_args="-std=c++11 -O3 -mcpu=native -Wno-unused-but-set-variable -Wno-sign-compare".split(),
++                extra_compile_args="-std=c++20 -O3 -mcpu=native -Wno-unused-but-set-variable -Wno-sign-compare".split(),
```
Comment (smuzaffar) [inline professor2-ppc64-flag-change.patch]: @iarspider , may be for cpp std we just patch the Makefile using perl/sed comamnd e.g.
```
#Make sure the default c++sdt stand is c++11 in Makefile
grep -q 'std=c++11' Makefile
#Change c++ std to %{cms_cxx_standard}
sed  -e 's|-std=c++11|-std=c++%{cms_cxx_standard}|' Makefile
```

File: scram-tools.file/tools/gcc/env.sh [hunk 2]
```diff
@@ -34,7 +34,7 @@ source ${SCRAM_TOOLS_BIN_DIR}/os_libdir.sh
 # optimizations as they become available in gcc.
 
 GCC_CXXFLAGS=""
-GCC_CXXFLAGS="$GCC_CXXFLAGS -std=c++20 -ftree-vectorize"
+GCC_CXXFLAGS="$GCC_CXXFLAGS -std=@CMS_CXX_STANDARD@ -ftree-vectorize"
```
Comment (smuzaffar) [inline scram-tools.file/tools/gcc/env.sh]: why not just use `c++$CMS_CXX_STANDARD` here?

File: scram-tools.file/bin/get_tools [hunk 3]
```diff
@@ -12,7 +12,10 @@ export SCRAM_TOOL_SOURCE_DIR=$(realpath ${SCRAM_TOOLS_BIN_DIR}/../tools)/$4
 
 #setup tools specific env
 tool_script=${SCRAM_TOOL_SOURCE_DIR}/env.sh
-if [ -f $tool_script ]; then source $tool_script; fi
+if [ -f $tool_script ]; then
+  perl -p -i -e 's|\@([^@]*)\@|$ENV{$1}|g' $tool_script
+  source $tool_script
+fi
```
Comment (smuzaffar) [inline scram-tools.file/bin/get_tools]: I think this is not needed if `scram-tools.file/tools/gcc/env.sh` uses `CMS_CXX_STANDARD` env directly

Comment (iarspider) [reply scram-tools.file/bin/get_tools]: I wasn't sure if this env variable will (still) be defined at that point. I will change this in the next iteration.

Comment (smuzaffar) [reply scram-tools.file/bin/get_tools]: it should be defined. Just like `COMPILER_CXXFLAGS`  which is set https://github.com/cms-sw/cmsdist/pull/8558/files#diff-9ae93e4d39c78dbb82f952da0248bf67d11ce362f51e329364779da933923564R12  and used by https://github.com/cms-sw/cmsdist/pull/8558/files#diff-09500a932b32fbe40198bc05cbf2ebbb173183365f9a4c751795eb25ac9c7ddeR62

**Comment by iarspider** [comment]
please test for CMSSW_13_2_CPP20_X

File: professor2.spec [hunk 4]
```diff
@@ -14,6 +15,11 @@ Patch0: professor2-ppc64-flag-change
 %patch0 -p1
 %endif
 
+#Make sure the default c++sdt stand is c++11 in Makefile
+grep -q 'std=c++11' pyext/setup.py
```
Comment (smuzaffar) [inline professor2.spec]: ah it is in `pyext/setup.py` , please update the comment :-)

File: scram-tools.file/tools/gcc/env.sh [hunk 5]
```diff
@@ -34,7 +34,7 @@ source ${SCRAM_TOOLS_BIN_DIR}/os_libdir.sh
 # optimizations as they become available in gcc.
 
 GCC_CXXFLAGS=""
-GCC_CXXFLAGS="$GCC_CXXFLAGS -std=c++20 -ftree-vectorize"
+GCC_CXXFLAGS="$GCC_CXXFLAGS -std=${CMS_CXX_STANDARD} -ftree-vectorize"
```
Comment (smuzaffar) [inline scram-tools.file/tools/gcc/env.sh]: I guess we are missing `c++` here i.e. it should be `-std=c++${CMS_CXX_STANDARD}` ... right?

**Comment by iarspider** [comment]
please test for CMSSW_13_2_CPP20_X

**Comment by iarspider** [comment]
Doing a local build, will retrigger the test once things are fixed.

**Comment by smuzaffar** [comment]
few more things to fix
- Fix the comment https://github.com/cms-sw/cmsdist/pull/8558/files#diff-b399077ccf647e6626255da9c774809c464ff5d77ef4b44407b487535911fdbfR18
- we also have cuda cpp stand in https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_13_2_X/g12_cpp20/cuda-flags.file , ay include `cpp-standard` in `cuda-flags.file ` and use
```
%define nvcc_stdcxx -std=c++%{cms_cxx_standard}
```

**Comment by smuzaffar** [comment]
@iarspider , why are we not moving forward with this?

**Comment by iarspider** [comment]
I am running a local test, will push once ready

**Comment by iarspider** [comment]
> I am running a local test, will push once ready

I also need to update TF in master first, then copy the changes here.

**Comment by iarspider** [comment]
test parameters:
- pull_request = cms-sw/cmssw#42228
- full_cmssw = true

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
please test

**Comment by smuzaffar** [comment]
abort
tests are running for el8_amd64_gcc11

**Comment by smuzaffar** [comment]
please test for CMSSW_13_3_CPP20_X

**Comment by iarspider** [comment]
please test for CMSSW_13_3_CPP20_X

**Comment by iarspider** [comment]
please test for CMSSW_13_3_CPP20_X

**Comment by iarspider** [comment]
The following packages failed:

* HeterogeneousCore/AlpakaInterface, HeterogeneousCore/AlpakaTest, HeterogeneousTest/CUDADevice, 
HeterogeneousTest/CUDAKernel, HeterogeneousTest/ROCmDevice, HeterogeneousTest/ROCmKernel, RecoTracker/PixelSeeding  - CUDA or ROCm compiler failure
* RecoLocalCalo/EcalRecProducers - https://github.com/cms-sw/cmssw/pull/42293

Once TF is merged into master, we can merge this as well.

**Comment by iarspider** [comment]
test parameters:
-  full_cmssw = true

**Comment by iarspider** [comment]
please test for CMSSW_13_3_CPP20_X

File: pip/requirements.txt [hunk 6]
```diff
@@ -214,16 +216,16 @@ notebook-shim==0.2.2
 numexpr==2.8.4
 # setuptools version <64 is needed by numpy: https://github.com/pypa/setuptools/issues/3549
 # NO_AUTO_UPDATE: update together with tensorflow
-numpy==1.23.5
+numpy==1.24.3
```
Comment (smuzaffar) [inline pip/requirements.txt]: does this work with current version of TF?

Comment (iarspider) [reply pip/requirements.txt]: With 2.6.4? IIRC, no.

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
please test for CMSSW_13_3_CPP20_X

**Comment by smuzaffar** [comment]
please test for CMSSW_13_3_CPP20_X

**Comment by smuzaffar** [comment]
lets merge it

**Pull request merged**